src/hugr/typecheck.rs

```diff
@@ -0,0 +1,163 @@
+    /// E.g. checking 300 against I8
+    IntTooLarge(u8, u128),
+    /// Width (n) of an `I<n>` type doesn't fit into a u32
+    IntWidthTooLarge(u8),
+    /// The width of an integer type wasn't a power of 2
+    IntWidthInvalid(u8),
+    /// Expected width (packed with const int) doesn't match type
+    IntWidthMismatch(u8, u8),
+    /// Found a Var type constructor when we're checking a const val
+    ConstCantBeVar,
+    /// The length of the tuple value doesn't match the length of the tuple type
+    TupleWrongLength,
+    /// Const values aren't allowed to be linear
+    LinearTypeDisallowed,
+    /// Tag for a sum value exceeded the number of variants
+    InvalidSumTag,
+    /// For a value which embeds its type (e.g. sum or opaque) - a mismatch
+    /// between the embedded type and the type we're checking against
+    TypeMismatch(ClassicType, ClassicType),
+    /// A mismatch between the embedded type and the type we're checking
+    /// against, as above, but for rows instead of simple types
+    TypeRowMismatch(TypeRow, TypeRow),
+}
+
+impl Display for TypeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let str = match self {
+            TypeError::Unimplemented(ty) => {
+                format!("Const type checking unimplemented for {}", ty)
+            }
+            TypeError::Failed(typ) => format!("Invalid const value for type {}", typ),
+            TypeError::IntTooLarge(width, val) => {
+                format!("Const int {} too large for type I{}", val, width)
+            }
+            TypeError::IntWidthTooLarge(w) => format!("Int type too large: I{}", w),
+            TypeError::IntWidthInvalid(w) => {
+                format!(
+                    "The int type I{} is invalid, because {} is not a power of 2",
+                    w, w
+                )
+            }
+            TypeError::IntWidthMismatch(exp, act) => format!(
+                "Type mismatch for int: expected I{}, but found I{}",
+                exp, act
+            ),
+            TypeError::ConstCantBeVar => "Type of a const value can't be Var".to_string(),
+            TypeError::TupleWrongLength => "Tuple of wrong length".to_string(),
+            TypeError::LinearTypeDisallowed => {
+                "Linear types not allowed in const nodes".to_string()
+            }
+            TypeError::InvalidSumTag => "Tag of Sum value is invalid".to_string(),
+            TypeError::TypeMismatch(exp, act) => {
+                format!("Type mismatch for const - expected {}, found {}", exp, act)
+            }
+            TypeError::TypeRowMismatch(exp, act) => {
+                format!("Type mismatch for const - expected {}, found {}", exp, act)
+            }
+        };
+        f.write_str(&str)
+    }
+}
+
+/// Per the spec, valid widths for integers are 2^n for all n in [0,7]
+fn check_valid_width(width: u8) -> Result<(), TypeError> {
+    if width > 128 {
```

Comment:
I would suggest adding a type alias for the width somewhere
`pub type HUGR_INT_WIDTH = u8;`

and then using things like 
```suggestion
    if width > (1<<HUGR_INT_WIDTH::BITS) {
```

Reply:
this should make it easier for us to change the width in future

Reply:
I've opted for `HugrIntWidthStore` and `HugrIntValueStore` because otherwise rust complains about non camel case type names